### PR TITLE
 Split term canonize utility to own file and document

### PR DESCRIPTION
**Pull request by ajreynol**
This is work towards #1216, and is in preparation for symmetry detection reusing the alpha equivalence module.

### CODE REVIEW

File: src/theory/quantifiers/term_canonize.cpp [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief Implementation of term canonize.
+ **/
+
+#include "theory/quantifiers/term_canonize.h"
+
+#include "theory/quantifiers/term_util.h"
+
+using namespace std;
```
Comment (HanielB) [inline src/theory/quantifiers/term_canonize.cpp]: kill

File: src/theory/quantifiers/term_canonize.cpp [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+  if (a.getKind() == BOUND_VARIABLE)
+  {
+    if (b.getKind() == BOUND_VARIABLE)
+    {
+      return a.getAttribute(InstVarNumAttribute())
+             < b.getAttribute(InstVarNumAttribute());
+    }
+    return true;
+  }
+  else if (b.getKind() != BOUND_VARIABLE)
```
Comment (HanielB) [inline src/theory/quantifiers/term_canonize.cpp]: no-else-after-return

File: src/theory/quantifiers/term_canonize.cpp [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+  {
+    Node aop = a.hasOperator() ? a.getOperator() : a;
+    Node bop = b.hasOperator() ? b.getOperator() : b;
+    Trace("aeq-debug2") << a << "...op..." << aop << std::endl;
+    Trace("aeq-debug2") << b << "...op..." << bop << std::endl;
+    if (aop == bop)
+    {
+      if (a.getNumChildren() == b.getNumChildren())
+      {
+        for (unsigned i = 0, size = a.getNumChildren(); i < size; i++)
```
Comment (HanielB) [inline src/theory/quantifiers/term_canonize.cpp]: cache size

File: src/theory/quantifiers/term_canonize.cpp [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+    std::stringstream oss;
+    oss << tn;
+    std::string typ_name = oss.str();
+    while (typ_name[0] == '(')
+    {
+      typ_name.erase(typ_name.begin());
+    }
+    std::stringstream os;
+    os << typ_name[0] << i;
+    Node x = NodeManager::currentNM()->mkBoundVar(os.str().c_str(), tn);
```
Comment (HanielB) [inline src/theory/quantifiers/term_canonize.cpp]: cache outside loop

File: src/theory/quantifiers/term_canonize.cpp [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+    TypeNode tn = n.getType();
+    // allocate variable
+    unsigned vn = var_count[tn];
+    var_count[tn]++;
+    Node fv = getCanonicalFreeVar(tn, vn);
+    visited[n] = fv;
+    Trace("canon-term-debug") << "...allocate variable." << std::endl;
+    return fv;
+  }
+  else if (n.getNumChildren() > 0)
```
Comment (HanielB) [inline src/theory/quantifiers/term_canonize.cpp]: no-else-after-return

File: src/theory/quantifiers/term_canonize.cpp [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+      std::sort(cchildren.begin(), cchildren.end(), sto);
+    }
+    // now make canonical
+    Trace("canon-term-debug") << "Make canonical children" << std::endl;
+    for (unsigned i = 0, size = cchildren.size(); i < size; i++)
+    {
+      cchildren[i] =
+          getCanonicalTerm(cchildren[i], apply_torder, var_count, visited);
+    }
+    if (n.getMetaKind() == kind::metakind::PARAMETERIZED)
```
Comment (HanielB) [inline src/theory/quantifiers/term_canonize.cpp]: you are in kind namespace, you could drop the prefix

File: src/theory/quantifiers/term_canonize.h [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+   */
+  bool getTermOrder(Node a, Node b);
+  /** get canonical free variable #i of type tn */
+  Node getCanonicalFreeVar(TypeNode tn, unsigned i);
+  /** get canonical term
+   *
+   * This returns a canonical (alpha-equivalent) version of n, where
+   * bound variables in n may be replaced by other ones, and arguments of
+   * commutative operators of n may be sorted (if apply_torder is true).
+   * In detail, we replace bound variables in n, right to left, so that
```
Comment (HanielB) [inline src/theory/quantifiers/term_canonize.h]: missing the rest of the doc

File: src/theory/quantifiers/term_canonize.cpp (hunk 4 above)
Comment (ajreynol) [reply src/theory/quantifiers/term_canonize.cpp]: This should be a new variable on every iteration of the loop.

File: src/theory/quantifiers/term_canonize.cpp (hunk 5 above)
Comment (ajreynol) [reply src/theory/quantifiers/term_canonize.cpp]: I'd prefer keeping this else, it takes up no space and helps distinguish this is another case of n.

File: src/theory/quantifiers/term_canonize.cpp (hunk 4 above)
Comment (HanielB) [reply src/theory/quantifiers/term_canonize.cpp]: the node manager, I mean

File: src/theory/quantifiers/term_canonize.cpp (hunk 3 above)
Comment (HanielB) [reply src/theory/quantifiers/term_canonize.cpp]: my bad, please ignore

**Pull request merged**